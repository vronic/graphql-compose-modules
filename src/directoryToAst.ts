import fs from 'fs';
import { join, resolve, dirname, basename } from 'path';

export interface DirectoryToAstOptions {
  relativePath?: string;
  extensions?: string[];
  include?: RegExp | ((path: string, kind: 'dir' | 'file', filename: string) => boolean);
  exclude?: RegExp | ((path: string, kind: 'dir' | 'file', filename: string) => boolean);
}

export type AstNodeKinds = 'rootType' | 'dir' | 'file' | 'root';

export interface AstBaseNode {
  kind: AstNodeKinds;
  name: string;
  absPath: string;
}

export interface AstRootTypeNode extends AstBaseNode {
  kind: 'rootType';
  children: AstDirChildren;
  namespaceConfig?: AstFileNode;
}

export type AstDirChildren = {
  [key: string]: AstDirNode | AstFileNode;
};

export interface AstDirNode extends AstBaseNode {
  kind: 'dir';
  children: AstDirChildren;
  namespaceConfig?: AstFileNode;
}

export interface AstFileNode extends AstBaseNode {
  kind: 'file';
  code: {
    default?: any;
  };
}

export type RootTypeNames = 'query' | 'mutation' | 'subscription';

export interface AstRootNode extends AstBaseNode {
  kind: 'root';
  children: {
    [T in RootTypeNames]?: AstRootTypeNode;
  };
}

export const defaultOptions: DirectoryToAstOptions = {
  extensions: ['js', 'ts'],
};

export function directoryToAst(
  m: NodeModule,
  options: DirectoryToAstOptions = defaultOptions
): AstRootNode {
  // if no path was passed in, assume the equivelant of __dirname from caller
  // otherwise, resolve path relative to the equivalent of __dirname
  const schemaPath = options?.relativePath
    ? resolve(dirname(m.filename), options.relativePath)
    : dirname(m.filename);

  // setup default options
  Object.keys(defaultOptions).forEach((prop) => {
    if (typeof (options as any)[prop] === 'undefined') {
      (options as any)[prop] = (defaultOptions as any)[prop];
    }
  });

  const result = {
    kind: 'root',
    name: basename(schemaPath),
    absPath: schemaPath,
    children: {},
  } as AstRootNode;

  fs.readdirSync(schemaPath).forEach((filename) => {
    const absPath = join(schemaPath, filename);

    if (fs.statSync(absPath).isDirectory()) {
      const dirName = filename;
      const re = /^(query|mutation|subscription)(\.(.*))?$/i;
      const found = dirName.match(re);
      if (found) {
        const opType = found[1].toLowerCase() as keyof AstRootNode['children'];
        let rootTypeAst = result.children[opType];
        if (!rootTypeAst)
          rootTypeAst = {
            kind: 'rootType',
            name: opType,
            absPath,
            children: {},
          } as AstRootTypeNode;

        const astDir = getAstForDir(m, absPath, options);
        if (astDir) {
          const subField = found[3]; // any part after dot (eg for `query.me` will be `me`)
          if (subField) {
            rootTypeAst.children[subField] = {
              ...astDir,
              name: subField,
              absPath,
            };
          } else {
            rootTypeAst.children = astDir.children;
            if (astDir.namespaceConfig) {
              rootTypeAst.namespaceConfig = astDir.namespaceConfig;
            }
          }
          result.children[opType] = rootTypeAst;
        }
      }
    }
  });

  return result;
}

export function getAstForDir(
  m: NodeModule,
  absPath: string,
  options: DirectoryToAstOptions = defaultOptions
): AstDirNode | void {
  const name = basename(absPath);

  if (!checkInclusion(absPath, 'dir', name, options)) return;

  const result: AstDirNode = {
    kind: 'dir',
    absPath,
    name,
    children: {},
  };

  // get the path of each file in specified directory, append to current tree node, recurse
  fs.readdirSync(absPath).forEach((filename) => {
    const absFilePath = join(absPath, filename);

    const stat = fs.statSync(absFilePath);
    if (stat.isDirectory()) {
      // this node is a directory; recurse
      if (result.children[filename]) {
        throw new Error(
          `You have a folder and file with same name "${filename}" by the following path ${absPath}. Please remove one of them.`
        );
      }
      const astDir = getAstForDir(m, absFilePath, options);
      if (astDir) {
        result.children[filename] = astDir;
      }
    } else if (stat.isFile()) {
      // this node is a file
      const fileAst = getAstForFile(m, absFilePath, options);
      if (fileAst) {
        if (fileAst.name === 'index') {
          result.namespaceConfig = fileAst;
        } else if (result.children[fileAst.name]) {
          throw new Error(
            `You have a folder and file with same name "${fileAst.name}" by the following path ${absPath}. Please remove one of them.`
          );
        } else {
          result.children[fileAst.name] = fileAst;
        }
      }
    }
  });

  return result;
}

export function getAstForFile(
  m: NodeModule,
  absPath: string,
  options: DirectoryToAstOptions = defaultOptions
): AstFileNode | void {
  const filename = basename(absPath);
  if (absPath !== m.filename && checkInclusion(absPath, 'file', filename, options)) {
    // module name shouldn't include file extension
    const moduleName = filename.substring(0, filename.lastIndexOf('.'));
    return {
      kind: 'file',
      name: moduleName,
      absPath,
      code: m.require(absPath),
    };
  }
}

function checkInclusion(
  absPath: string,
  kind: 'dir' | 'file',
  filename: string,
  options: DirectoryToAstOptions
): boolean {
  // Skip dir/files started from double underscore
  if (/^__.*/i.test(filename)) {
    return false;
  }

  // Skip dir/files started from dot
  if (/^\..*/i.test(filename)) {
    return false;
  }

  if (kind === 'file') {
    if (
      // Verify file has valid extension
      !new RegExp('\\.(' + (options?.extensions || ['js', 'ts']).join('|') + ')$', 'i').test(
        filename
      ) ||
      // Hardcoded skip file extensions
      // typescript definition files
      new RegExp('(\\.d\\.ts)$', 'i').test(filename) ||
      // test files
      new RegExp('(\\.spec\\.(ts|js))$', 'i').test(filename)
    ) {
      return false;
    }
  }

  if (options.include) {
    if (options.include instanceof RegExp) {
      // if options.include is a RegExp, evaluate it and make sure the path passes
      if (!options.include.test(absPath)) return false;
    } else if (typeof options.include === 'function') {
      // if options.include is a function, evaluate it and make sure the path passes
      if (!options.include(absPath, kind, filename)) return false;
    }
  }

  if (options.exclude) {
    if (options.exclude instanceof RegExp) {
      // if options.exclude is a RegExp, evaluate it and make sure the path doesn't pass
      if (options.exclude.test(absPath)) return false;
    } else if (typeof options.exclude === 'function') {
      // if options.exclude is a function, evaluate it and make sure the path doesn't pass
      if (options.exclude(absPath, kind, filename)) return false;
    }
  }

  return true;
}
