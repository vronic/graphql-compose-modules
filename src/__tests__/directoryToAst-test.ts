import { directoryToAst } from '../directoryToAst';

describe('directoryToAst()', () => {
  describe('Schema ./__testSchema__', () => {
    const ast = directoryToAst(module, { relativePath: './__testSchema__' });

    it('should return root types', () => {
      expect(Object.keys(ast.children)).toEqual(expect.arrayContaining(['query', 'mutation']));
    });

    it('query has proper values', () => {
      expect(ast.children.query).toMatchSnapshot({
        name: 'query',
        kind: 'rootType',
        absPath: expect.any(String),
        children: {
          auth: {
            kind: 'dir',
            name: 'auth',
            absPath: expect.any(String),
            children: {
              isLoggedIn: expect.objectContaining({ kind: 'file' }),
              nested: expect.objectContaining({ kind: 'dir' }),
            },
          },
          field: expect.objectContaining({ kind: 'file' }),
          me: {
            absPath: expect.any(String),
            children: {
              'address.city': expect.objectContaining({ kind: 'file' }),
              'address.street': expect.objectContaining({ kind: 'file' }),
              name: expect.objectContaining({ kind: 'file' }),
            },
            namespaceConfig: {
              kind: 'file',
              name: 'index',
              absPath: expect.any(String),
              code: expect.any(Object),
            },
          },
          'some.nested': expect.objectContaining({ kind: 'file' }),
          'some.index': expect.any(Object),
          user: {
            absPath: expect.any(String),
            children: {
              extendedData: expect.any(Object),
              roles: expect.any(Object),
            },
            namespaceConfig: expect.any(Object),
          },
        },
        namespaceConfig: expect.any(Object),
      });
    });

    it('mutation has proper values', () => {
      expect(ast.children.mutation).toMatchSnapshot({
        name: 'mutation',
        kind: 'rootType',
        absPath: expect.any(String),
        children: {
          auth: {
            absPath: expect.any(String),
            children: {
              login: expect.objectContaining({ kind: 'file' }),
              logout: expect.objectContaining({ kind: 'file' }),
              nested: expect.objectContaining({ kind: 'dir' }),
            },
            namespaceConfig: expect.objectContaining({ kind: 'file' }),
          },
          user: {
            absPath: expect.any(String),
            children: {
              create: expect.objectContaining({ kind: 'file' }),
              update: expect.objectContaining({ kind: 'file' }),
            },
          },
          'logs.nested': {
            absPath: expect.any(String),
            children: {
              list: expect.objectContaining({ kind: 'file' }),
            },
          },
        },
      });
    });
  });
});
