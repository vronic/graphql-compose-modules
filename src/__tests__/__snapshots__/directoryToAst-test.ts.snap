// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directoryToAst() Schema ./__testSchema__ mutation has proper values 1`] = `
Object {
  "absPath": Any<String>,
  "children": Object {
    "auth": Object {
      "absPath": Any<String>,
      "children": Object {
        "login": ObjectContaining {
          "kind": "file",
        },
        "logout": ObjectContaining {
          "kind": "file",
        },
        "nested": ObjectContaining {
          "kind": "dir",
        },
      },
      "kind": "dir",
      "name": "auth",
      "namespaceConfig": ObjectContaining {
        "kind": "file",
      },
    },
    "logs.nested": Object {
      "absPath": Any<String>,
      "children": Object {
        "list": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "logs.nested",
    },
    "user": Object {
      "absPath": Any<String>,
      "children": Object {
        "create": ObjectContaining {
          "kind": "file",
        },
        "update": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "user",
    },
  },
  "kind": "rootType",
  "name": "mutation",
}
`;

exports[`directoryToAst() Schema ./__testSchema__ query has proper values 1`] = `
Object {
  "absPath": Any<String>,
  "children": Object {
    "auth": Object {
      "absPath": Any<String>,
      "children": Object {
        "isLoggedIn": ObjectContaining {
          "kind": "file",
        },
        "nested": ObjectContaining {
          "kind": "dir",
        },
      },
      "kind": "dir",
      "name": "auth",
      "namespaceConfig": Object {
        "absPath": "/Users/nodkz/www/_npm/graphql-compose-modules/src/__tests__/__testSchema__/query.auth/index.ts",
        "code": Object {
          "default": Object {
            "type": "NamespaceCustomTypeName",
          },
        },
        "kind": "file",
        "name": "index",
      },
    },
    "field": ObjectContaining {
      "kind": "file",
    },
    "me": Object {
      "absPath": Any<String>,
      "children": Object {
        "address.city": ObjectContaining {
          "kind": "file",
        },
        "address.street": ObjectContaining {
          "kind": "file",
        },
        "name": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "me",
      "namespaceConfig": Object {
        "absPath": Any<String>,
        "code": Any<Object>,
        "kind": "file",
        "name": "index",
      },
    },
    "some.index": Any<Object>,
    "some.nested": ObjectContaining {
      "kind": "file",
    },
    "user": Object {
      "absPath": Any<String>,
      "children": Object {
        "extendedData": Any<Object>,
        "roles": Any<Object>,
      },
      "kind": "dir",
      "name": "user",
      "namespaceConfig": Any<Object>,
    },
  },
  "kind": "rootType",
  "name": "query",
  "namespaceConfig": Any<Object>,
}
`;
